Add CMake support

kreuzerkrieg opened the pull request:
Add support for cmake build system.
Why?
Works on all platforms/architectures.
No need to get stuck with ancient VS sln files
Generates stuff for later consuming the output library in your projects (cpuidConfig.cmake stuff)

Actually, I need `libcpuid` as `vcpkg` and it will be a lot easier to create vcpkg port if the build system is cmake, thats why :)

anrieff commented on CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+set(VERSION "0.4.1")
+project(cpuid LANGUAGES C CXX ASM VERSION ${VERSION})
+
+set(CMAKE_CXX_STANDARD 20)
```
Would this request C++20 support and fail building on systems which don't have recent enough compiler? I'm not much versed with CMake, hence the question.
If that is indeed the case, we don't need such restrictions; C++11 and C99 are enough for libcpuid.

kreuzerkrieg replied:
Good point, will change rightaway

kreuzerkrieg replied:
Pushed.

kreuzerkrieg commented:
> Otherwise looks good, but I'm not really the person to review CMake configuration.

Even if there are atrocities in cmake since the change is not intrusive it wouldnt harm in any way

anrieff commented:
Yup, thanks for the clarification. Many thanks for your time!

X0rg commented:
@kreuzerkrieg On Linux, the CMake install was broken and incomplete. I fixed that in 7179a7b103495b9a46d5926d42aef86585ebcc7a and I added missing targets. Let me know if you observe regressions on Windows build.

kreuzerkrieg commented:
> @kreuzerkrieg On Linux, the CMake install was broken and incomplete. I fixed that in [7179a7b](https://github.com/anrieff/libcpuid/commit/7179a7b103495b9a46d5926d42aef86585ebcc7a) and I added missing targets. Let me know if you observe regressions on Windows build.

not sure what do you mean by "broken and incomplete" but, by all means go ahead

as of Windows, I cant observe regression on windows since I dont build on Win

X0rg commented:
By "broken and incomplete", here was the list of installed files I got:
```
/usr/lib/libcpuid.a
/usr/lib/cmake/cpuid/cpuidConfig.cmake
/usr/lib/cmake/cpuid/cpuidConfigVersion.cmake
/usr/lib/cmake/cpuid/cpuidTargets.cmake
/usr/lib/cmake/cpuid/cpuidTargets-debug.cmake
/usr/bin/Debug/cpuid_tool
```
`/usr/bin/Debug/cpuid_tool` is not a valid path for a binary.

Now, since my changes:
```shell
/usr/lib/pkgconfig/libcpuid.pc
/usr/lib/libcpuid.so.0.4.1
/usr/lib/libcpuid.so.14
/usr/lib/libcpuid.so
/usr/include/libcpuid.h
/usr/include/libcpuid_types.h
/usr/include/libcpuid_constants.h
/usr/lib/cmake/cpuid/cpuidConfig.cmake
/usr/lib/cmake/cpuid/cpuidConfigVersion.cmake
/usr/share/man/man3/cpu_sgx_t.3
/usr/share/man/man3/cpu_raw_data_t.3
/usr/share/man/man3/cpu_mark_t.3
/usr/share/man/man3/cpu_list_t.3
/usr/share/man/man3/cpu_id_t.3
/usr/share/man/man3/cpu_epc_t.3
/usr/share/man/man3/libcpuid.3
/usr/share/man/man3/cpuid_tool.3
/usr/bin/cpuid_tool
```
It installs development files and man pages.

kreuzerkrieg commented:
> By "broken and incomplete", here was the list of installed files I got:
> 
> ```
> /usr/lib/libcpuid.a
> /usr/lib/cmake/cpuid/cpuidConfig.cmake
> /usr/lib/cmake/cpuid/cpuidConfigVersion.cmake
> /usr/lib/cmake/cpuid/cpuidTargets.cmake
> /usr/lib/cmake/cpuid/cpuidTargets-debug.cmake
> /usr/bin/Debug/cpuid_tool
> ```
> 
> `/usr/bin/Debug/cpuid_tool` is not a valid path for a binary.
> 
> Now, since my changes:
> 
> ```shell
> /usr/lib/pkgconfig/libcpuid.pc
> /usr/lib/libcpuid.so.0.4.1
> /usr/lib/libcpuid.so.14
> /usr/lib/libcpuid.so
> /usr/include/libcpuid.h
> /usr/include/libcpuid_types.h
> /usr/include/libcpuid_constants.h
> /usr/lib/cmake/cpuid/cpuidConfig.cmake
> /usr/lib/cmake/cpuid/cpuidConfigVersion.cmake
> /usr/share/man/man3/cpu_sgx_t.3
> /usr/share/man/man3/cpu_raw_data_t.3
> /usr/share/man/man3/cpu_mark_t.3
> /usr/share/man/man3/cpu_list_t.3
> /usr/share/man/man3/cpu_id_t.3
> /usr/share/man/man3/cpu_epc_t.3
> /usr/share/man/man3/libcpuid.3
> /usr/share/man/man3/cpuid_tool.3
> /usr/bin/cpuid_tool
> ```
> 
> It installs development files and man pages.

ah, this... 
actually, it was never intended to be "installed" except via `vcpkg` as stated in the beginning of this PR. The question is if it will keep working in `vcpkg` once you have merged your changes. 
Once merged, ping me, I will check if `vcpkg` still works.

X0rg commented:
> actually, it was never intended to be "installed" except via `vcpkg`

Ah, ok, never mind, I'm unfamiliar with `vcpkg`.

>  Once merged, ping me, I will check if vcpkg still works.
These changes are already on master (I see I forgot to add `cpuidTargets` for Unix system, will be fixed soon).

Yes, if you can do a test to ensure I have not broken anything with `vcpkg`. :thinking:

kreuzerkrieg commented:
> > actually, it was never intended to be "installed" except via `vcpkg`
> 
> Ah, ok, never mind, I'm unfamiliar with `vcpkg`.
> 
> > Once merged, ping me, I will check if vcpkg still works.
> > These changes are already on master (I see I forgot to add `cpuidTargets` for Unix system, will be fixed soon).
> 
> Yes, if you can do a test to ensure I have not broken anything with `vcpkg`.

ok, looks like it is broken since you do not install target's export. so one cant just `find_package(cpuid REQUIRED)` in his cmake.
Please advise.

X0rg commented:
Hmm, I see: `vcpkg` is not included in `WIN32`... Fixed in c129bc23a4db3f29141ba321aa2de59c0f4a4121. Sorry for the inconvenience.

kreuzerkrieg commented:
> Hmm, I see: `vcpkg` is not included in `WIN32`... Fixed in [c129bc2](https://github.com/anrieff/libcpuid/commit/c129bc23a4db3f29141ba321aa2de59c0f4a4121). Sorry for the inconvenience.

No headers for me.
I think there is a problem with `INSTALL(... INCLUDES PUBLIC_HEADER DESTINATION...)`. I think it should be or `INCLUDES DESTINATION` or `PUBLIC_HEADER DESTINATION`. Did you get it working the way it is written now?

https://cmake.org/cmake/help/latest/command/install.html

kreuzerkrieg commented:
hm... looks like, physically it is presented, but the target does not add `INTERFACE_INCLUDE_DIRECTORIES` to the target.

X0rg commented:
@kreuzerkrieg Ok, so I ended by doing some tests with `vcpkg`:
```shell
# vcpkg install --head cpuid
Computing installation plan...
The following packages will be built and installed:
    cpuid[core]:x64-linux (from HEAD)
Warning: abi keys are missing values:
    head

Starting package 1/1: cpuid:x64-linux
Building package cpuid[core]:x64-linux from HEAD...
-- Purging cached /var/cache/vcpkg/anrieff-libcpuid-master.tar.gz to fetch latest (use --no-downloads to suppress)
-- Downloading https://api.github.com/repos/anrieff/libcpuid/git/refs/heads/master...
-- Skipping hash check for /var/cache/vcpkg/temp/anrieff-libcpuid-master.tar.gz.version.
-- Downloading https://github.com/anrieff/libcpuid/archive/master.tar.gz...
-- Skipping hash check for /var/cache/vcpkg/temp/anrieff-libcpuid-master.tar.gz.
-- Extracting source /var/cache/vcpkg/anrieff-libcpuid-master.tar.gz
-- Applying patch fix-install-headers.patch
-- Using source at /usr/share/vcpkg/buildtrees/cpuid/src/head/master-31e2e8bb01
-- Configuring x64-linux-dbg
-- Configuring x64-linux-rel
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Installing: /usr/share/vcpkg/packages/cpuid_x64-linux/share/cpuid/copyright
-- Performing post-build validation
-- Performing post-build validation done
Building package cpuid[core]:x64-linux... done
Installing package cpuid[core]:x64-linux...
Installing package cpuid[core]:x64-linux... done
Elapsed time for package cpuid:x64-linux: 2.402 s

Total elapsed time: 2.402 s

The package cpuid:x64-linux provides CMake targets:

    find_package(cpuid CONFIG REQUIRED)
    target_link_libraries(main PRIVATE cpuid::cpuid)
```
I'm able to build CPU-X with Libcpuid installed by `vcpkg` by using `/usr/share/vcpkg/scripts/buildsystems/vcpkg.cmake` toolchain (plus appropriate changes in CPU-X `CMakeLists.txt`).

Can you confirm ac5702b06b3220932b83b7654f4ad987ce08cf3d fix your last issue? I use the directive `#include <libcpuid/libcpuid.h>` in my C code.

kreuzerkrieg commented:
Confirmed.
Everything worked as expected.
Thanks
